src/runtime/vulkan_context.h

```diff
@@ -0,0 +1,329 @@
+#ifndef HALIDE_RUNTIME_VULKAN_CONTEXT_H
+#define HALIDE_RUNTIME_VULKAN_CONTEXT_H
+
+#include "printer.h"
+#include "runtime_internal.h"
+#include "scoped_spin_lock.h"
+
+#include "vulkan_extensions.h"
+#include "vulkan_internal.h"
+#include "vulkan_memory.h"
+
+// --------------------------------------------------------------------------
+
+namespace Halide {
+namespace Runtime {
+namespace Internal {
+namespace Vulkan {
+
+// --------------------------------------------------------------------------
+
+// An Vulkan context/queue/synchronization lock defined in this module with weak linkage
+// Vulkan Memory allocator for host-device allocations
+halide_vulkan_memory_allocator *WEAK cached_allocator = nullptr;
+VkInstance WEAK cached_instance = nullptr;
+VkDevice WEAK cached_device = nullptr;
+VkCommandPool WEAK cached_command_pool = 0;
+VkQueue WEAK cached_queue = nullptr;
+VkPhysicalDevice WEAK cached_physical_device = nullptr;
+uint32_t WEAK cached_queue_family_index = 0;
+volatile ScopedSpinLock::AtomicFlag WEAK thread_lock = 0;
+
+// --------------------------------------------------------------------------
+
+// Helper object to acquire and release the Vulkan context.
+class VulkanContext {
+    void *user_context;
+
+public:
+    VulkanMemoryAllocator *allocator;
+    VkInstance instance;
+    VkDevice device;
+    VkCommandPool command_pool;
+    VkPhysicalDevice physical_device;
+    VkQueue queue;
+    uint32_t queue_family_index;  // used for operations requiring queue family
+    VkResult error;
+
+    HALIDE_ALWAYS_INLINE VulkanContext(void *user_context)
+        : user_context(user_context),
+          allocator(nullptr),
+          instance(nullptr),
+          device(nullptr),
+          command_pool(0),
+          physical_device(nullptr),
+          queue(nullptr),
+          queue_family_index(0),
+          error(VK_SUCCESS) {
+
+        int result = halide_vulkan_acquire_context(user_context,
+                                                   reinterpret_cast<halide_vulkan_memory_allocator **>(&allocator),
+                                                   &instance, &device, &physical_device, &command_pool, &queue, &queue_family_index);
+        halide_abort_if_false(user_context, result == 0);
```

Comment:
(1) is `halide_abort_if_false()` really the best way to deal with errors here? (ie, is it unrecoverable and `abort()` the only option?) 
(2) do we really need to check all the values being output (in addition to the function result)? for a debug-runtime-only build I'd be fine with this, but this feels like overkill for a normal runtime.

Reply:
Agreed ... this was just a quick check that I added for testing ... it definitely needs to be more robust. I'll fix this.